src/expr/src/expr/expr_regexp_count.rs

```diff
@@ -0,0 +1,252 @@
+            bail!("syntax error in `regexp_count`");
+        }
+
+        let flags = if let Some(f) = flags {
+            f
+        } else {
+            "".to_string()
+        };
+
+        if flags.contains('g') {
+            bail!("`regexp_count` does not support global flag option");
+        }
+
+        let ctx = RegexpContext::new(&pattern, &flags)?;
+
+        Ok(Self {
+            source,
+            ctx,
+            return_type,
+            start,
+        })
+    }
+}
+
+impl RegexpCountExpression {
+    fn match_row(&self, text: Option<&str>) -> Option<i32> {
+        if let Some(text) = text {
+            // First get the start position to count for
+            let start = if let Some(s) = self.start { s - 1 } else { 0 };
+
+            // For unicode purpose
+            let mut start = match text.char_indices().nth(start as usize) {
+                Some((idx, _)) => idx,
+                // The start is out of bound or just invalid
+                // FIXME: What to return?
+                None => return None,
+            };
+
+            let mut count = 0;
+
+            while let Some(captures) = self.ctx.regex.captures(&text[start..]) {
+                count += 1;
+                start += captures.get(0).unwrap().end();
+            }
+
+            Some(count)
+        } else {
+            // Input string is None
+            // FIXME: Return None or Some(0)?
+            None
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for RegexpCountExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let source_column = self.source.eval_checked(input).await?;
+        let source_column = source_column.as_utf8();
+
+        let row_len = input.capacity();
+        let vis = input.vis();
+        let mut builder: I32ArrayBuilder = ArrayBuilder::new(row_len);
+
+        for row_idx in 0..row_len {
+            if !vis.is_set(row_idx) {
+                builder.append_n(1, None);
+                continue;
+            }
+
+            let source = match source_column.value_at(row_idx) {
+                Some(s) => s,
+                None => {
+                    builder.append_n(1, None);
+                    continue;
+                }
+            };
+
+            if let Some(ret) = self.match_row(Some(source)) {
+                builder.append_n(1, Some(ret));
+            } else {
+                builder.append_n(1, None);
+            }
+        }
+
+        Ok(Arc::new(ArrayImpl::from(builder.finish())))
+    }
+
+    async fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
+        let source = self.source.eval_row(input).await?;
+        let source = match source {
+            Some(ScalarImpl::Utf8(s)) => s,
+            _ => return Ok(None),
+        };
+
+        Ok(self
+            .match_row(Some(&source))
+            .map(|replaced| replaced.into()))
```

Comment:
should panic if the source is not a string.
```suggestion
        let source = source.as_utf8();
        Ok(self
            .match_row(source)
            .map(|replaced| replaced.into()))
```

Reply:
Seems that we could not directly use `as_utf8` to convert a `Option<ScalarImpl>` to `Option<&str>`🤔️?
And that's why I use the match to deserialize the `source`.

If the expected behavior is to panic, I think the following changes will satisfy this purpose.
```
let source = match source {
    Some(ScalarImpl::Utf8(s)) => s,
    _ => bail!("source should be a String"),
};
```

Reply:
Oh, what about `source.map(|s| s.as_utf8())`? Does that work?

Reply:
Your approach should also work. But need to add a branch: `None => return Ok(None),`.

Reply:
I see, I update the version that sticks with the `match`, since the `source.map(|s| s.as_utf8())` is returning the reference, which is not allowed due to the borrow checker😭.
Rather than doing the type exercise at the moment, I prefer keeping it simple first🤣.